### DISCUSSION TITLE
The future of system CRDs and how they're made available to workspaces

### DISCUSSION THREAD

**Discussion by ncdc** [kcp-dev/kcp#980] [Development]
We will reserve (although it's not documented) `kcp.dev` CRDs for kcp system use (when https://github.com/kcp-dev/kcp/issues/772 is implemented). We put all the system CRDs in the `system:system-crds` logical cluster. To avoid copying them into all the other logical clusters that use them, we have custom CRD handling logic to make system CRDs available in certain types of workspaces. Which workspaces see which system CRDs is currently hard-coded:

https://github.com/kcp-dev/kcp/blob/f4f4a73effe2a94fc98e5d6609aefa6f15f82eaf/pkg/server/apiextensions.go#L75-L99

This is not going to scale, so we'll presumably want to make this more flexible and configurable. Some thoughts on how to do this:

1. Consider making only the APIExport/APIBinding-related CRDs available to every workspace. Any additional types would be injected via individual APIBindings per workspace.
2. Remove the hard-coded list and move it into ClusterWorkspace.Spec.
3. Some combination of the above

WDYT? Other ideas? Don't bother? 😄 Would love to hear opinions.

**Comment by sttts** [comment]
Would love to see (1). We could place the APIExport for that in some system workspace. The interesting bit: then those need a shared identity between different shards such that in the future (when we have wildcard authz for serviceaccount with identities as audience) service account can access these kcp-wide.

**Reply by ncdc** [reply]
Does (1) increase the "cost" of a workspace because it now needs 1..n APIBinding CRs, replicated times thousands of workspaces?

**Reply by sttts** [reply]
true. We could have system bindings, that are magically applied, but don't exist as objects :-) dejavu.

**Reply by ncdc** [reply]
We could also move the hard-coding into spec fields on ClusterWorkspace or something. Not a great solution, but at least not in go code.
